job-server/src/spark.jobserver/AkkaClusterSupervisorActor.scala

```diff
@@ -0,0 +1,217 @@
+
+  private val contexts = mutable.HashMap.empty[String, ActorRef]
+  private val resultActors = mutable.HashMap.empty[String, ActorRef]
+
+  private val cluster = Cluster(context.system)
+  private val selfAddress = cluster.selfAddress
+
+  // This is for capturing results for ad-hoc jobs. Otherwise when ad-hoc job dies, resultActor also dies,
+  // and there is no way to retrieve results.
+  val globalResultActor = context.actorOf(Props[JobResultActor], "global-result-actor")
+
+  logger.info("AkkaClusterSupervisor initialized on {}", selfAddress)
+
+  override def preStart(): Unit = {
+    cluster.join(selfAddress)
+    cluster.subscribe(self, initialStateMode = InitialStateAsEvents, classOf[MemberEvent])
+  }
+
+  override def postStop(): Unit = {
+    cluster.unsubscribe(self)
+  }
+
+  def wrappedReceive: Receive = {
+    case MemberUp(member) =>
+      if (member.hasRole("manager")) {
+        val memberActors = RootActorPath(member.address) / "user" / "*"
+        context.actorSelection(memberActors) ! Identify(memberActors)
+      }
+
+    case ActorIdentity(memberActors, actorRefOpt) =>
+      actorRefOpt.map { actorRef =>
+        val actorName = actorRef.path.name
+        if (actorName.startsWith("jobManager")) {
+          logger.info("Received identify response, attempting to initialize context at {}", memberActors)
+          val initInfoOpt = contextInitInfos.remove(actorName)
+          val callbackOpt = contextCallbacks.remove(actorName)
+          if (initInfoOpt.isDefined && callbackOpt.isDefined) {
+            val (ctxName, ctxConf, isAdHoc) = initInfoOpt.get
+            val (successFunc, failureFunc) = callbackOpt.get
+            initContext(actorName, actorRef, contextInitTimeout)(
+              ctxName, ctxConf, isAdHoc)(successFunc, failureFunc)
+          }
+          else {
+            logger.warn("No initialization or callback found for jobManager actor {}", actorRef.path)
+            actorRef ! PoisonPill
+          }
+        }
+      }
+
+
+    case AddContextsFromConfig =>
+      addContextsFromConfig(config)
+
+    case ListContexts =>
+      sender ! contexts.keys.toSeq
+
+    case AddContext(name, contextConfig) =>
+      val originator = sender()
+      val mergedConfig = contextConfig.withFallback(defaultContextConfig)
+      if (contexts contains name) {
+        originator ! ContextAlreadyExists
+      } else {
+        startContext(name, mergedConfig, false) { ref =>
+          originator ! ContextInitialized
+        } { err =>
+          originator ! ContextInitError(err)
+        }
+      }
+
+    case GetAdHocContext(classPath, contextConfig) =>
+      val originator = sender()
+      val mergedConfig = contextConfig.withFallback(defaultContextConfig)
+
+      var contextName = ""
+      do {
+        contextName = java.util.UUID.randomUUID().toString().substring(0, 8) + "-" + classPath
+      } while (contexts contains contextName)
```

Comment:
This will try to check that we don't have 2 contexts with the same name, but we set the value only when the context has been initialized, that leaves a lot of time for other requests to create the same context.